File: platform/processing-core/app/schemas/admin/policies.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel


class PolicyType(str, Enum):
    FLEET = "fleet"
    FINANCE = "finance"
    MARKETPLACE = "marketplace"


class PolicyStatus(str, Enum):
    ENABLED = "enabled"
    DISABLED = "disabled"


class PolicyScope(BaseModel):
    tenant_id: int | None = None
    client_id: str | None = None


class PolicyExplainRef(BaseModel):
    kind: str
    id: str
    type: PolicyType


class AdminPolicyIndexItem(BaseModel):
    id: str
    type: PolicyType
    title: str
    status: PolicyStatus
    scope: PolicyScope
    actions: list[str]
    explain_ref: PolicyExplainRef
    updated_at: datetime | None = None
    toggle_supported: bool = False


class AdminPolicyIndexResponse(BaseModel):
    items: list[AdminPolicyIndexItem]
    total: int
    limit: int
    offset: int


class AdminPolicyHeader(BaseModel):
    id: str
    type: PolicyType
    title: str
    status: PolicyStatus
    scope: PolicyScope
    actions: list[str]
    updated_at: datetime | None = None
    toggle_supported: bool = False


class AdminPolicyDetailResponse(BaseModel):
    header: AdminPolicyHeader
    policy: dict[str, Any] | None
    explain: dict[str, Any] | None = None


class AdminPolicyExecutionOut(BaseModel):
    id: str
    policy_id: str
    event_type: str
    event_id: str
    action: str
    status: str
    reason: str | None = None
    created_at: datetime


class AdminPolicyExecutionListResponse(BaseModel):
    items: list[AdminPolicyExecutionOut]

File: frontends/admin-ui/src/api/policies.ts
import { request } from "./http";

export type PolicyType = "fleet" | "finance" | "marketplace";
export type PolicyStatus = "enabled" | "disabled";

export interface PolicyScope {
  tenant_id: number | null;
  client_id: string | null;
}

export interface PolicyExplainRef {
  kind: string;
  id: string;
  type: PolicyType;
}

export interface PolicyIndexItem {
  id: string;
  type: PolicyType;
  title: string;
  status: PolicyStatus;
  scope: PolicyScope;
  actions: string[];
  explain_ref: PolicyExplainRef;
  updated_at?: string | null;
  toggle_supported: boolean;
}

export interface PolicyIndexResponse {
  items: PolicyIndexItem[];
  total: number;
  limit: number;
  offset: number;
}

export interface PolicyHeader {
  id: string;
  type: PolicyType;
  title: string;
  status: PolicyStatus;
  scope: PolicyScope;
  actions: string[];
  updated_at?: string | null;
  toggle_supported: boolean;
}

export interface PolicyDetailResponse {
  header: PolicyHeader;
  policy: Record<string, unknown> | null;
  explain?: Record<string, unknown> | null;
}

export interface PolicyExecution {
  id: string;
  policy_id: string;
  event_type: string;
  event_id: string;
  action: string;
  status: string;
  reason?: string | null;
  created_at: string;
}

export interface PolicyExecutionResponse {
  items: PolicyExecution[];
}

export async function listPolicies(
  token: string,
  params: { type?: PolicyType; status?: PolicyStatus; q?: string; limit?: number; offset?: number },
): Promise<PolicyIndexResponse> {
  const query = new URLSearchParams();
  if (params.type) query.set("type", params.type);
  if (params.status) query.set("status", params.status);
  if (params.q) query.set("q", params.q);
  if (params.limit) query.set("limit", String(params.limit));
  if (params.offset) query.set("offset", String(params.offset));
  const suffix = query.toString();
  return request<PolicyIndexResponse>(`/policies${suffix ? `?${suffix}` : ""}`, { method: "GET" }, { token });
}

export async function fetchPolicyDetail(token: string, type: PolicyType, id: string): Promise<PolicyDetailResponse> {
  return request<PolicyDetailResponse>(`/policies/${type}/${id}`, { method: "GET" }, { token });
}

export async function fetchPolicyExecutions(
  token: string,
  type: PolicyType,
  id: string,
): Promise<PolicyExecutionResponse> {
  return request<PolicyExecutionResponse>(`/policies/${type}/${id}/executions`, { method: "GET" }, { token });
}

export async function enablePolicy(token: string, type: PolicyType, id: string): Promise<PolicyHeader> {
  return request<PolicyHeader>(`/policies/${type}/${id}/enable`, { method: "POST" }, { token });
}

export async function disablePolicy(token: string, type: PolicyType, id: string): Promise<PolicyHeader> {
  return request<PolicyHeader>(`/policies/${type}/${id}/disable`, { method: "POST" }, { token });
}
